naber's stuff

alemreM opened the pull request:
wym cant automatically merge

Z3rca commented:
merged with agreement of hunharibo

hunharibo committed:
fix inventory use button, re-enabled ink mechanics

alemreM committed:
wip changes from testing-and-stuff

chaos/herdstone reward (1,3 - 2,4) is still bugged for sturgia/aserai
dot spell lag is still
didnt have to fix brawl behavior :)

alemreM committed:
comments applied

hunharibo committed:
fix enchanment screen icons

hunharibo committed:
fix firearms

hunharibo committed:
add back artillery, some wonky behavior with reloading, formations and detached agents

alemreM committed:
things

fixed decimals
fixed balanced and changed orc boss quests
removed legendary bows from loot pools
fixed catalyst perk
fixed waaagh unit cap
fixed pool drops

alemreM committed:
oops

dlls are overrated anyway

Zedforen committed:
First set of dwarf banners

- Hirn
- Kadrin
- Izor

alemreM committed:
Merge branch 'development' into nabers-playhouse

SlyDevil committed:
Dismemberment, cursed site, skills, and text fixes. Native companion behavior replaced.

Fixed dismemberment crash in friendly missions.
Companion behavior is now active, but there are some minor fixes that are not yet implemented.
Fixed crash when hovering mouse over a cursed site. The decal has been made more visible to help with uneven terrain near the settlement.
Fixed skills granting too large of bonuses due to them not being adjusted as percentages by the game anymore.
Added a text line to the shader caching menu option describing how a user knows when it's complete.
Added a late patch and removed SubModule initialization of GameText manager.
Added a xslt to delete all of the native wanderers.
Fixed town fortification levels being at 2 instead of 3 for 3/4 of the Towns - this is purely for settlement construction purposes; due to how scene levels are loaded, the intended scene was always loaded.

Added some placeholder text lines for trainer npcs.

Zedforen committed:
Second set of dwarf banners

- Angazhar
- Azgaraz
- Gantuk
- Norn
- Zhufbar
- Ziflin

hunharibo committed:
fix settlement nameplates for good

alemreM committed:
crash fix and miscast testing

alemreM committed:
Revert "crash fix and miscast testing"

This reverts commit 718540984dfcaea947256f0ddc28b151baaab3f7.

alemreM committed:
crash fix and miscast test

alemreM committed:
Merge branch 'nabers-playhouse' of https://github.com/TheOldRealms/TOR_Core into nabers-playhouse

alemreM committed:
Merge branch 'development' into nabers-playhouse

alemreM committed:
merged dlls

alemreM committed:
Merge branch 'development' into nabers-playhouse

alemreM committed:
merged dlls

Zedforen committed:
Final dwarf banners

- Eksfilaz
- Kaferkammaz

Z3rca committed:
specialization icons vampires

Pierogolepca committed:
Character creator and text changes

Fixed a bunch of text in strings
Added remaining lore texts for CC screen 4
Added vampire skill/attribute bonuses for CC screen 4
(will add specific CC gear sets later)

Z3RCA MOVE THE PANEL UP PLS AAAAAAAAAAARGGH

alemreM committed:
Merge branch 'development' into nabers-playhouse

alemreM committed:
Merge branch 'development' into nabers-playhouse

alemreM committed:
Merge branch 'development' into nabers-playhouse

alemreM committed:
Merge branch 'development' into nabers-playhouse

alemreM committed:
Merge branch 'development' into nabers-playhouse

Z3rca committed:
changing  SpellCraft to Spellcraft

alemreM committed:
SpellCraft

Z3rca committed:
temp fix for crashing game (fixed with shaman)

SlyDevil committed:
Wanderer spawning fully functional. Campaign time updated to use the new model. Combat xp not penalized in certain contexts.

Old wanderers correctly removed and unregistered.
Fixed a trebuchet crash and poked at artillery ammo issues.
Fixed a conversation loop with empire magister skill training.
Fixed rare bug when a save occurs just after a player loses a battle.
Added time model : restored old start time and changed nighttime to be the 8 hour period between 20h-4h.
Adjusted combat xp to always use the 1x multiplier as if in a field battle unless explicitly set to grant no xp in a context.
Updated athel loren check for enchantment ingredient drop calculation.
Vampire party speed bonuses adjusted to make use of the new nighttime period.

alemreM committed:
changes on vanilla medicine

not tested

alemreM committed:
removed miscast so zerca wont get frustrated

alemreM committed:
how could i forget

SlyDevil commented on CSharpSourceCode/AbilitySystem/AbilityManagerMissionLogic.cs:
```diff
@@ -722,40 +723,64 @@ private void SetUpCastStanceParticles()
+                    continue;
+
+                var hero = element.Character.HeroObject;
+                if (hero == null)
```
Character.IsHero looks for a non-null HeroObject already, no?

party.GetMemberHeroes can return a list of heroes and performs the same checks if you want to use the existing extension

SlyDevil commented on CSharpSourceCode/CampaignMechanics/Crafting/EnchantmentIngredientLootCampaignBehavior.cs:
```diff
@@ -58,26 +56,60 @@ public override void SyncData(IDataStore dataStore)
+        );
 
-        foreach (var pair in _goodAmounts)
+        var targetRoster = PlayerEncounter.Current?.RosterToReceiveLootItems ?? PartyBase.MainParty.ItemRoster;
```
I'd toss in an info message that tells the player that the items were added directly to their inventory here so they don't feel like they missed out on something if the encounter ends up null for whatever reason, but that they still received loot from it.

Or perhaps logging with TORCommon.Log so we have a message about when/how the first option could be null. Could be removed after beta testing so any performance concern is only present during testing.

SlyDevil commented on CSharpSourceCode/CampaignMechanics/Crafting/EnchantmentIngredientLootCampaignBehavior.cs:
```diff
@@ -58,26 +56,60 @@ public override void SyncData(IDataStore dataStore)
-            var amount = model.CalculateResultAmount(pair.Value, pair.Key, playerEarnedLootPercentage);
-            var item = TorEnchantingIngredients.GetItemObjectForIngredient(pair.Key);
-            itemRosterToReceive.Add(new ItemRosterElement(item, amount));
+            var factorSum = _goodAmounts[ingredientType];
```
Can you rename _goodAmounts to _goodsFactors? I think the prior usage dates back to the original writing by zerca which may be different than what it's now used for.

SlyDevil commented on CSharpSourceCode/CampaignMechanics/CustomResourceBehavior/TeefBehavior.cs:
```diff
@@ -361,6 +313,32 @@ private void AddGoldForTeef(List<InquiryElement> inquiryElements)
+            if (item == null)
+                continue;
+
+            if (item.StringId != null && item.StringId.StartsWith("tor_gs_"))
```
Are there items that have ids starting with "tor_gs_" that aren't marked as Culture.aserai?

This can probably be checked more directly via (item.Culture == TORConstants.Cultures.GREENSKINS).

SlyDevil commented on CSharpSourceCode/CampaignMechanics/CustomResourceBehavior/TeefBehavior.cs:
```diff
@@ -151,63 +151,15 @@ void CreateTeefContainer(List<InquiryElement> inquiryElements)
+            InventoryScreenHelper.OpenScreenAsReceiveItems(
+                donatedItems,
+                new TextObject("Give Items to the Big boss"),
+                () => OnItemsDiscarded(donatedItems));
```
much cleaner, very nice

SlyDevil commented on CSharpSourceCode/CampaignMechanics/CustomResources/WaaaghMeter/WaaaghBehavior.cs:
```diff
@@ -49,39 +54,56 @@ private void CalculateWaaaghGainFromBattle(MapEvent mapEvent)
 
         var playerWon = mapEvent.WinningSide == mapEvent.PlayerSide;
 
-        if (playerWon)
```
Keep the playerWon check here.

If they lost, the delta can be set immediately without bothering to calculate the ratio and other information that will have no bearing.

SlyDevil commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
```diff
@@ -114,58 +116,129 @@ public List<string> GetUnlockedOakUpgradeCategory(string unlockedUpgradeCategory
+                var rewardItems = comp.RewardItemIds
+                    .Select(id => MBObjectManager.Instance.GetObject<ItemObject>(id))
+                    .Where(i => i != null)
+                    .Where(i => _xmlItemIds.Contains(i.StringId))
```
I'm not so sure about this check.

_xmlItemIds is filled via OnSessionLaunched which will trigger for both new and saved games. Does this event precede the save game being read and any additional item ids not present in the xml (ie. enchanted items generated previously and baked into the save) having already been added to the ObjectManager for its "Items" list?
Have you verified that enchanted items gained aren't present in the hash set?

If they are, then specificItems will include them and the player may be offered a magic item they already have (as items not in their inventory are periodically deleted).

SlyDevil commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
```diff
@@ -114,58 +116,129 @@ public List<string> GetUnlockedOakUpgradeCategory(string unlockedUpgradeCategory
                 {
-                    itemIds = itemIds.TakeRandom(2).ToList();
-                }
+                    var rewardSelection = rewardItems;
```
This logic creates a weird boundary condition when the raid component only contains 1 or 2 items in its list.

I'm also confused by the random selection, then filtering for the recipes to learn, rather than selecting from just the recipes with some chance scaled by rewardItems.Count.
Is the goal in separating out the recipes to dilute the equipment items from the drop list with the cultural items while maintaining the drop rates on the recipes?

Recipes can probably be filtered directly by checking the item's type is "Goods".

SlyDevil commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
```diff
@@ -114,58 +116,129 @@ public List<string> GetUnlockedOakUpgradeCategory(string unlockedUpgradeCategory
+                var specificItems = comp.RewardItemIds
+                    .Select(id => MBObjectManager.Instance.GetObject<ItemObject>(id))
+                    .Where(i => i != null)
+                    .Where(i => _xmlItemIds.Contains(i.StringId))
```
If the item exists in the object manager, and has been listed in the drop list, then it must necessarily exist in the hashset because that was created from the ObjectManager.

Rather than just null checking here, the drop lists should be verified on campaign initialization and listed items that don't yet exist in the weapon xmls should be logged to preemptively keep track of problems.

The TorItem check should be superfluous; if we've added native equipment to our drop lists, that's our problem to take care of at a different point rather than rechecking here.

Similarly, CraftedByPlayer and artifacts should not be added to the drop list by us.

The culture check is confusing. Much of the drop table will be filtered out by it.

SlyDevil commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
```diff
@@ -114,58 +116,129 @@ public List<string> GetUnlockedOakUpgradeCategory(string unlockedUpgradeCategory
+                    .Where(i => i != null)
+                    .Where(i => _xmlItemIds.Contains(i.StringId))
+                    .Where(i => i.IsTorItem()) // no vanilla
+                    .Where(i => (i.IsWeapon() || i.IsArmor()))
```
If someone wants a shield to drop from the raid, this will prevent that from ever dropping with no indication as to why despite the expectation of the drop list being the "intended items".

SlyDevil commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
```diff
@@ -114,58 +116,129 @@ public List<string> GetUnlockedOakUpgradeCategory(string unlockedUpgradeCategory
 
                 var model = (TORBattleRewardModel)Campaign.Current.Models.BattleRewardModel;
 
                 var newItems = new List<ItemObject>(items);
```
Why not take the ~4 items from equipmentItems, modify them with traits, and add them directly to the final reward list and avoid having to copy book+equipment to items, items to newItems, filtering items to not add traits to the book, then adding the modified equipment piece to newItems and removing the unmodified version from newItems?

SlyDevil commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
```diff
@@ -359,6 +443,10 @@ private void OnSessionLaunched(CampaignGameStarter starter)
 
         CollectSettlementData();
+        if (_xmlItemIds == null)
+            _xmlItemIds = MBObjectManager.Instance.GetObjectTypeList<ItemObject>()
```
Has this been checked for player crafted items to verify that no enchanted items have been loaded into the manager at this point?

SlyDevil commented on CSharpSourceCode/HarmonyPatches/MedicineSurgeryChancePatch.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace TOR_Core.HarmonyPatches
+{
+    [HarmonyPatch(typeof(DefaultPartyHealingModel), "GetSurvivalChance")]
```
Was there a reason to do this via a transpile rather than our existing override for that method?

SlyDevil reviewed (commented):
Looks good. A couple things that stand out.

Not sure why it picked up so many changes from other commits and bloated the files changed list.

alemreM committed:
Merge branch 'development' into nabers-playhouse

alemreM committed:
dlls

alemreM committed:
pr comments

applied the comments and some additional polish. campaign map event trait influences have been quadrupled. not tested.

SlyDevil commented on CSharpSourceCode/CampaignMechanics/Crafting/EnchantmentIngredientLootCampaignBehavior.cs:
```diff
@@ -92,21 +92,30 @@ private void SetLootedIngredients(MapEvent mapEvent)
+        if (usedLootFallback && anyIngredientAdded)
+        {
+            InformationManager.DisplayMessage(new InformationMessage(
+                new TextObject("{=tor_crafting_ingredient_loot_fallback}Looted enchantment ingredients were added directly to your party inventory.").ToString()));
```
str_tor_crafting_ingredient_loot_fallback

string id starts with "tor", localization id starts with "str_tor"

SlyDevil commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
```diff
@@ -221,15 +249,24 @@ private void OnMissionEnded(IMission obj)
                 foreach (var item in items)
                 {
-                    if (!item.IsWeapon() && (!item.IsArmor() || item.HasAnyTrait()))
+                    var canReceiveTraits = item.IsWeapon() || (item.IsArmor() && !item.HasAnyTrait());
```
`(item.IsWeapon() || item.IsArmor()) && !item.HasAnyTrait()`

Current parentheses check for IsWeapon OR (IsArmor and the armor has no traits), but that feels off to me.

Looking at the current code before your changes and I'm just as confused by the original logic; why would a weapon that already has a trait have more added, but an armor with a trait can't have any more added.
`ItemTrait.IsValidFor` has a case to cover shield itemtype, and the incoming items were already filtered to weapon/armor/shield, so this bool can simply be `!item.HasAnyTrait()`.

For consistency between weapons and armor+shield, I find the simplest to simply exclude anything with a pre-existing trait from receiving more which is what I wrote above.
There's a potential to add parity by checking the difference between `item.GetTraits().Count` and traitCount, to then add the difference of traits to the item regardless of type, but I'd leave that out for now and ask others to weigh in on preferences for a future change.

SlyDevil commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
```diff
@@ -239,10 +276,12 @@ private void OnMissionEnded(IMission obj)
                 }
-                items = newItems;
+
+                unmodifiedItems.AddRange(enchantedItems);
```
Add `blueprintRewardItems` as well; recipes are currently not added even if they were selected as a drop further up.

unmodifiedItems comes from equipmentItems which comes from selectedItems/cultureItems, but none of these prior lists had the recipes added to them and specifically filtered non-equipment out.

SlyDevil replied:
I see what you mentioned with the blueprints getting added previously; I missed the foreach loop that handles it.

SlyDevil commented on ModuleData/tor_strings.xml:
```diff
@@ -2796,7 +2804,68 @@
+	<string id="tor_spelltrainer_greenskin_goodbye_str" text="{=tor_spelltrainer_greenskin_goodbye_str}Get outta 'ere! You'ze dispersin' my smoke." />
+
+	<!--Vanilla perk changes-->
+    <string id="GJRcFc0V" text="{=GJRcFc0V}A variable chance to recover a mount from lost cavalry after battles." />
```
I'm unsure how the gametextmanager will store/fetch this in relation to the pre-existing string defined in DefaultsPerks.

Does this modified string show correctly in english and another arbitrary language (should also show the english string in that language)?

If yes, fine by me for the moment.

SlyDevil reviewed (approved):
Final comments for minor changes.
I'm curious what the drops will look like from raids, but there's no immediate problem.

alemreM committed:
comments + fixes

main hero encyclopedia page crash fixed
Veterinarian overrride fixed
previous comments applied

SlyDevil commented on CSharpSourceCode/HarmonyPatches/VeterinarianPatch.cs:
```diff
@@ -53,4 +57,52 @@ private static float GetMountRecoveryChance(ItemObject mountItem)
+    {
+        private const string VETERINARIAN_PERK_ID = "MedicineVeterinarian";
+        private const string SECONDARY_DESCRIPTION_REPLACEMENT =
+            "{=tor_veterinarian_description}A variable chance to recover a mount from lost cavalry after battles.";
```
str_tor_veterinarian_description

Edit : I'll handle this and adding it to strings xml after merge.

SlyDevil reviewed (approved):
Last comment for the localization id.

Can merge after.

Z3rca commented on CSharpSourceCode/Quests/Careers/OrcBossQuest2.cs:
```diff
@@ -15,13 +15,15 @@ namespace TOR_Core.Quests.Careers
         private const int RequiredCitiesCaptured = 5;
         private const int RequiredLordDuels = 15;
-        private const int RequiredTeefTransferred = 500000;
+        private const int RequiredTeefTransferred = 1000000;
```
i would review this with gamedesign

Z3rca commented on CSharpSourceCode/Quests/Careers/OrcBossQuest1.cs:
```diff
@@ -163,9 +163,29 @@ private void OnSkillIncreased(Hero hero, SkillObject skill, int skillValueBefore
-            _currentBattlesWon++;
-            _taskBattlesWon.UpdateCurrentProgress(_currentBattlesWon);
-            UpdateQuest();
+            if (_currentBattlesWon == 1)
```
there is a bug i see it but what does this do?

Z3rca commented on CSharpSourceCode/Quests/Careers/OrcBossQuest1.cs:
```diff
@@ -14,11 +14,11 @@ namespace TOR_Core.Quests.Careers
-        private const int RequiredArenaFights = 20;
-        private const int RequiredTeefTransferred = 150000;
-        private const int RequiredBrawlsWon = 50;
+        private const int RequiredWeaponSkillLevels = 175;
```
again gamedesign debate.

Z3rca commented on CSharpSourceCode/CampaignMechanics/TORCustomSettlement/TORCustomSettlementCampaignBehavior.cs:
i would like to have a bit more description here, what is your intention, for this? this seems to work but has this been discussed with game design?

emrebahat replied:
excluded artifacts/legendary bows, it was a request
blueprints are now explicitly added into the *items*, so other checks won't work on them, i believe this was the reason why their drop rate felt incredibly low
after you get any blueprints, he number of equipment picks is reduced to keep the total roughly constant. this could (and i think should) be adjusted furthermore.
older culture check was bugged, also didn't exclude crafted items/vanilla items for overriden cultures.
old code had weapons that could still get more traits even if already traited, shields weren't included and armor couldn't if traited
tier checks has been added so you are less likely to end up with some crap of your culture
iirc my pr log also mentioned them. if you want me to add comments to a specific part of the code or any of these changes to be discussed with game design lmk.

Z3rca reviewed (changes requested):
overall I dont see much issues, rather than it requires more detailed explanation WHY those changes are made. I dont see a description in the pr, nor some comments of yours inside the code why balance changes were made. please act on these , maybe consult piero. I would require you to check this with designers before it gets approved.

alemreM committed:
fixes and reverted changes

reverted orcboss quests, fixed strings, a temp patch for inventory reset bug

alemreM committed:
you didn't see that
